server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/join/merge/MergeSortKeyComparator.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.execution.operator.process.join.merge;
+
+import org.apache.iotdb.db.utils.datastructure.MergeSortKey;
+import org.apache.iotdb.tsfile.utils.Pair;
+
+import java.io.Serializable;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class MergeSortKeyComparator implements Comparator<MergeSortKey>, Serializable {
+
+  private final boolean nullFirst;
+  private final int index;
+  private final Comparator<MergeSortKey> originalComparator;
+  private static final Map<
+          Pair<Comparator<MergeSortKey>, Pair<Boolean, Integer>>, MergeSortKeyComparator>
+      comparatorCache = new ConcurrentHashMap<>();
+
+  public MergeSortKeyComparator(
+      int index, boolean nullFirst, Comparator<MergeSortKey> originalComparator) {
+    this.nullFirst = nullFirst;
+    this.index = index;
+    this.originalComparator = originalComparator;
+  }
+
+  public static MergeSortKeyComparator getInstance(
+      int index, boolean nullFirst, Comparator<MergeSortKey> originalComparator) {
+    return comparatorCache.computeIfAbsent(
+        new Pair<>(originalComparator, new Pair<>(nullFirst, index)),
+        comparator -> new MergeSortKeyComparator(index, nullFirst, originalComparator));
```

Comment:
This cache doesn't take effective at all. Each time the originalComparator won't be the same.